java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java

```diff
@@ -0,0 +1,301 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.closeresource;
+
+import org.sonar.java.checks.SubscriptionBaseVisitor;
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.java.checks.methods.MethodInvocationMatcherCollection;
+import org.sonar.java.checks.methods.TypeCriteria;
+import org.sonar.java.symexecengine.DataFlowVisitor;
+import org.sonar.java.symexecengine.State;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeCastTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+public class CloseableVisitor extends DataFlowVisitor {
+
+  private static final MethodInvocationMatcherCollection CLOSE_INVOCATIONS = closeMethodInvocationMatcher();
+
+  private static final String[] IGNORED_CLOSEABLE_SUBTYPES = {
+    "java.io.ByteArrayOutputStream",
+    "java.io.ByteArrayInputStream",
+    "java.io.StringReader",
+    "java.io.StringWriter",
+    "java.io.CharArraReader",
+    "java.io.CharArrayWriter"
+  };
+
+  private static final String CLOSE_METHOD_NAME = "close";
+  private static final String JAVA_IO_CLOSEABLE = "java.io.Closeable";
+  private static final String JAVA_LANG_AUTOCLOSEABLE = "java.lang.AutoCloseable";
+
+  public CloseableVisitor(List<VariableTree> methodParameters, SubscriptionBaseVisitor check) {
+    super(check);
+    ignoreVariables(methodParameters);
+  }
+
+  private void ignoreVariables(List<VariableTree> variables) {
+    for (VariableTree methodParameter : variables) {
+      super.visitVariable(methodParameter);
+      ignoreVariable(methodParameter);
+    }
+  }
+
+  private void ignoreVariable(VariableTree variableTree) {
+    Symbol symbol = variableTree.symbol();
+    if (isCloseableOrAutoCloseableSubtype(symbol.type())) {
+      executionState.markValueAs(symbol, new CloseableState.Ignored(variableTree));
```

Comment:
It is useless to test before marking as ignored, as only the symbol which are already known could be marked. Note that to be known the symbol have to pass the "isSymbolRelevant" test, which is equivalent.